File: src/classes/RangeHelper.js
import _ from "lodash";

export default class RangeHelper {
    ranges = [];

    constructor(ranges = []) {
        if (!_.isEmpty(ranges)) {
            this.addRanges(ranges);
        }
    }

    addRanges(ranges) {
        for (let range of ranges) {
            this.addRange(range.from, range.to, range.value);
        }
    }

    addRange(from, to, value) {
        if (!_.isUndefined(this.getValue(from)) || !_.isUndefined(this.getValue(to))) {
            throw new Error("Ranges cannot overlap");
        }

        this.ranges.push({
            from: from,
            to: to,
            value: value
        });
    }

    getValue(count) {
        for (let range of this.ranges) {
            if (count >= range.from && count <= range.to) {
                return range.value;
            }
        }

        return undefined;
    }
};

File: bot/src/Utility/RangeHelper.cs
using System;
using System.Collections.Generic;

namespace Shinobu.Utility
{
    public class RangeHelper<T>
    {
        private List<Range<T>> _ranges = new List<Range<T>>();
        
        public RangeHelper(Range<T>[] ranges)
        {
            foreach (var range in ranges)
            {
                AddRange(range);
            }
        }

        public void AddRange(Range<T> range)
        {
            if (null != GetValue(range.From) || (null != range.To && null != GetValue((int) range.To)))
            {
                throw new Exception("Ranges cannot overlap");
            }

            _ranges.Add(range);
        }

        public T? GetValue(int value)
        {
            foreach (var r in _ranges)
            {
                if (value >= r.From && (r.To == null || value <= r.To))
                {
                    return r.Value;
                }
            }

            return default(T);
        }
    }
}